pytest_freezegun.py

```diff
@@ -1,18 +1,52 @@
 # -*- coding: utf-8 -*-
 
+import time
+
+import _pytest.runner
 import pytest
+
 from freezegun import freeze_time
 
 
-@pytest.yield_fixture
+@pytest.fixture
 def freezer(request):
-    marker = request.node.get_marker('freeze_time')
-    args, kwargs = (marker.args, marker.kwargs) if marker else ((), {})
-    with freeze_time(*marker.args, **marker.kwargs) as freezer:
-        yield freezer
+    freezer, frozen_time = FreezegunPlugin.freezers.get(
+        request.node.nodeid,
+        (None, None)
+    )
+
+    if freezer is None:
+        with freeze_time() as frozen_time:
+            yield frozen_time
+    else:
+        yield frozen_time
+
+
+def patch_pytest_runner_time():
+    # Patch the time function in pytest's runner,
+    # so it reports the correct setup/call/teardown durations.
+    _pytest.runner.time = time.time.previous_time_function
+
+
+class FreezegunPlugin(object):
+    freezers = {}
+
+    @pytest.hookimpl(tryfirst=True)
+    def pytest_runtest_setup(self, item):
+        marker = item.get_marker('freeze_time')
+
+        if marker:
+            freezer = freeze_time(*marker.args, **marker.kwargs)
+            self.freezers[item.nodeid] = freezer, freezer.start()
```

Comment:
Why is `.freezers` a dict? Will there be multiple active freezers at a time?

Reply:
I had a feeling this would not work when running in parallel (using `pytest-xdist`), but have tested it now and it works fine as a regular class attribute. I thought it was running in threads, but it seems it's using separate processes, so we don't need to worry about thread safety.